tokio-codec/src/lines_codec.rs

```diff
@@ -41,19 +110,90 @@ impl Decoder for LinesCodec {
     type Error = io::Error;
 
     fn decode(&mut self, buf: &mut BytesMut) -> Result<Option<String>, io::Error> {
-        if let Some(newline_offset) =
-            buf[self.next_index..].iter().position(|b| *b == b'\n')
-        {
-            let newline_index = newline_offset + self.next_index;
-            let line = buf.split_to(newline_index + 1);
-            let line = &line[..line.len()-1];
-            let line = without_carriage_return(line);
-            let line = utf8(line)?;
-            self.next_index = 0;
-            Ok(Some(line.to_string()))
-        } else {
-            self.next_index = buf.len();
-            Ok(None)
+        let mut trim_and_offset = None;
+        loop {
```

Comment:
I'd like to explore a simpler implementation. My gut is that there should be a much simpler way to go about this.

For example, why not limit the buffer to max line when searching for the new line character? If it is not there, then transition to "discarding".

```rust
if is_discarding {
    // just discard to new line
} else {
    let newline_offset = buf[self.next_index..self.max_length].iter().position(....);

    if let Some(pos) = newline_offset {
        // success
    } else if buf.len() > self.max_length {
        // transition to discarding
    } else {
        // wait for more data.
    }
}
```

Reply:
@carllerche I tried rewriting this logic to be closer to your example in b2d237252f565c9860e79ed1f23dcc71c83df2cd. 

It turns out that there are still some edge cases that need to be handled, so the final code ended up being a bit more complex than what you sketched out, but I added a great deal of comments explaining all these cases in 0a078ca7130a2189ca40eb8d5c369735441feee5. Hopefully the new code is at least somewhat clearer than my prior attempt...